Neutrals flux limiters

malamast opened the pull request:
Just a few suggestions for neutral_mixed and other components:

1) added a functionality to initialize n and p from grid variables. 
2) Changed the way we calculate the flux limiters and transport coefficients of neutrals. I separate the gradients for the flux limiters in parallel and perpendicular directions. They do something similar in uedge as well. 
3) Added the contribution of ion velocities to the neutral perpendicular velocity. This term comes for momentum sources in the derivation of the neutral perpendicular velocity. How can we generalize it here for multiple species?
4) minor changes to zero_current and braginskii_collisions components.

**MK edits:**

I am taking the PR over from Malamas and finishing it. Tasks:
- [x] Changes from Malamas: separate limiters for parallel/perpendicular conductivity and viscosity, perpendicular ion velocity coupling, and so on
- [x] Change limiter formulation to AFN style
- [x] Add user settable limiters for each flux
- [x] Change free streaming velocities to AFN definition
- [x] Remove double-counting of `neutral_lmax`
- [x] Make perpendicular ion velocity coupling a user option
- [x] Test against historical SOLPS comparison results
- [ ] Update 2D-Production test
- [ ] Add flux limiter reconstruction to 2D-recycling and rename to 2D-reconstruct
- [ ] Documentation

malamast committed:
evolve_density and pressure: Added a functionality to initialize number density and pressure from grid variables with names varName_init.

fixed_density and isothermal: Added a functionality to read the fixed denstiy or temperature from grid variables with names varName_init

malamast committed:
zero_current: Set species momentum and added momentum to outputVars

malamast committed:
anomalous_diffusion: make anomalous_D variable accessible to other components. This willbe needed in neutral_mixed for calculating the contribution of the ion velocity.

malamast committed:
neutral_mixed: Added the contribution of ion velocity to the neutral perpendicular velocity to make it consistent with AFN models.

malamast committed:
neutral_mixed: I separated the way we calculate the flux limiters in the perpendicular and parallel directions. This is what they do in UEDGE as well.

-Added the contibution of ion velocities to the neural diffusion perpendicular velocity. This term comes form momentum source terms due to collisions of neutrals  with ions. See eq 20 amd 21 of Horsten 2017.

malamast committed:
braginskii_collisions: added a user-defined density_floor

mikekryjak commented:
I'm happy to deal with it - I need this to finish my ST40 comparison and paper.

bendudson committed:
BOUT-dev: Use branch with SNES solver updates

bendudson committed:
BOUT-dev: More SNES changes

Added jacobian_persists option

mikekryjak committed:
Fix rebase

mikekryjak committed:
Add new diagnostics

mikekryjak committed:
Improve Dnn logic flow

mikekryjak committed:
Add temporary debug variable

mikekryjak committed:
Fix segfault/wasted allocation

mikekryjak committed:
Add diagnostics to allow reproduction in test

mikekryjak committed:
Refactor eta for consistency with kappa

mikekryjak committed:
Improve diagnostic naming

mikekryjak committed:
[bot] Add last format changes commit to ignore file

mikekryjak committed:
[bot] Apply format changes

mikekryjak committed:
Formatting

mikekryjak committed:
Make double counting lmax a flag, refactor
neutral_lmax was already used in Dnn, it shouldn't be also put into Dmax as this limits it twice! The flag enables legacy behaviour by default.

mikekryjak committed:
Improve formulation of thermal speed
Left a flag for legacy behaviour.

mikekryjak committed:
Add catch for boolean flux_limit setting
This was boolean before. Now it's the numerical setting for the limiter.

mikekryjak committed:
Separate neutral limiters
With intuitive defaults.

mikekryjak committed:
Formatting

mikekryjak committed:
Remove spurious neutral_lmax re-definition

mikekryjak committed:
Remove flux limit exception duplicate due to poor merge

mikekryjak committed:
Implement AFN style limiter formulation
Unlike the previous form, there is a sharpness setting. This is referred to as gamma in the AFN parlance.

mikekryjak committed:
Add setting for perp ion velocity coupling
Previously always on

mikekryjak committed:
Formatting

mikekryjak committed:
Merge branch 'master' into neutrals-flux-limmiters

mikekryjak committed:
Merge remote-tracking branch 'origin/snes-controllers' into neutrals-flux-limmiters

mikekryjak committed:
Merge branch 'master' into neutrals-flux-limmiters

mikekryjak committed:
Merge branch 'master' into neutrals-flux-limmiters

mikekryjak committed:
Fix permissions

mikekryjak committed:
Implement improved D/kappa/eta max denominator regularisation

Based on https://github.com/boutproject/hermes-3/pull/605

mikekryjak committed:
Fix viscosity floor

Was mistakenly set to be the same as the other two floors, which is incorrect as e.g. it has different units. I had good results with 1e-8 which doesn't get engaged in steady state.
